pkg/innerring/internal/blockchain/blockchain.go

```diff
@@ -0,0 +1,500 @@
+	cfgBaseApp.P2P.AttemptConnPeers = int(cfg.P2P.AttemptConnPeers)
+	cfgBaseApp.P2P.MaxPeers = int(cfg.P2P.MaxPeers)
+
+	err = cfgBase.ProtocolConfiguration.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid composed protocol configuration: %w", err)
+	}
+
+	cfgServer, err := network.NewServerConfig(cfgBase)
+	if err != nil {
+		return nil, fmt.Errorf("compose NeoGo server config from the base one: %w", err)
+	}
+
+	var cfgDB dbconfig.DBConfiguration
+	cfgDB.Type = cfg.Storage.typ
+	if cfgDB.Type == dbconfig.BoltDB {
+		cfgDB.BoltDBOptions.FilePath = cfg.Storage.path
+	} else if cfgDB.Type == dbconfig.LevelDB {
+		cfgDB.LevelDBOptions.DataDirectoryPath = cfg.Storage.path
+	}
+
+	bcStorage, err := storage.NewStore(cfgDB)
+	if err != nil {
+		return nil, fmt.Errorf("init storage for blockchain: %w", err)
+	}
+
+	bc, err := core.NewBlockchain(bcStorage, cfgBase.Blockchain(), cfg.Logger)
+	if err != nil {
+		closeErr := bcStorage.Close()
+		if closeErr != nil {
+			return nil, fmt.Errorf("init core blockchain component: %w; failed to close blockchain storage: %v", err, closeErr)
+		}
+
+		return nil, fmt.Errorf("init core blockchain component: %w", err)
+	}
+
+	netServer, err := network.NewServer(cfgServer, bc, bc.GetStateSyncModule(), cfg.Logger)
+	if err != nil {
+		return nil, fmt.Errorf("init NeoGo network server: %w", err)
+	}
+
+	if !cfg.NotaryDisabled {
+		var cfgNotary notary.Config
+		cfgNotary.MainCfg.Enabled = true
+		cfgNotary.MainCfg.UnlockWallet = cfg.Wallet
+		cfgNotary.Chain = bc
+		cfgNotary.Log = cfg.Logger
+
+		notaryService, err := notary.NewNotary(cfgNotary, netServer.Net, netServer.GetNotaryPool(), func(tx *transaction.Transaction) error {
+			err := netServer.RelayTxn(tx)
+			if err != nil && !errors.Is(err, core.ErrAlreadyExists) {
+				return fmt.Errorf("relay completed notary transaction %s: %w", tx.Hash().StringLE(), err)
+			}
+
+			return nil
+		})
+		if err != nil {
+			return nil, fmt.Errorf("init Notary service: %w", err)
+		}
+
+		netServer.AddService(notaryService)
+		bc.SetNotary(notaryService)
+	}
+
+	var cfgConsensus consensus.Config
+	cfgConsensus.Logger = cfg.Logger
+	cfgConsensus.Broadcast = netServer.BroadcastExtensible
+	cfgConsensus.Chain = bc
+	cfgConsensus.BlockQueue = netServer.GetBlockQueue()
+	cfgConsensus.ProtocolConfiguration = bc.GetConfig().ProtocolConfiguration
+	cfgConsensus.RequestTx = netServer.RequestTx
+	cfgConsensus.StopTxFlow = netServer.StopTxFlow
+	cfgConsensus.TimePerBlock = cfg.BlockInterval
+	cfgConsensus.Wallet = cfg.Wallet
+
+	consensusService, err := consensus.NewService(cfgConsensus)
+	if err != nil {
+		return nil, fmt.Errorf("init Consensus service: %w", err)
+	}
+
+	netServer.AddConsensusService(consensusService, consensusService.OnPayload, consensusService.OnTransaction)
+
+	// "make" channel rw to satisfy Start method
+	chErrRw := make(chan error)
+
+	go func() {
+		err, ok := <-chErrRw
+		if ok {
+			cfg.ErrorListener <- err
+		}
+	}()
```

Comment:
This function will exit immediately. It must be the loop for services errors habdling. And it's something that doesn't work properly in neo-go as far, we have https://github.com/nspcc-dev/neo-go/issues/2896.

Reply:
Added loop